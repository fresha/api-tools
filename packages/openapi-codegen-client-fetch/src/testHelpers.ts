// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';

import type { ActionContext, Context } from './context';
import type { OpenAPIModel, OperationModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel): Context => {
  const base = createTSProjectTestContext(openapi);

  return {
    ...base,
    includeDeprecated: false,
    includedTags: new Set<string>(),
    excludedTags: new Set<string>(),
    transformRequest: 'kebab',
    transformResponse: 'camel',
  };
};

export const createActionTestContext = (
  operation: OperationModel,
  fileName: string,
): ActionContext => {
  const base = createTestContext(operation.root);
  const sourceFile = base.project.createSourceFile(fileName, '');

  return {
    ...base,
    sourceFile,
    operation,
  };
};
