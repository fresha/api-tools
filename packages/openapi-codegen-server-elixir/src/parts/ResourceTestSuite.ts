import assert from 'assert';

import { faker } from '@faker-js/faker';
import { JSONValue, snakeCase } from '@fresha/api-tools-core';

import type { Context } from './types';
import type { SourceFile } from '@fresha/code-morph-ex';
import type { ResourceModel } from '@fresha/json-api-model';

export class ResourceTestSuite {
  protected readonly context: Context;
  readonly moduleName: string;
  protected readonly resourceModuleName: string;
  protected readonly resourceModuleAlias: string;
  protected readonly resource: ResourceModel;
  protected readonly sourceFile: SourceFile;

  constructor(
    context: Context,
    moduleName: string,
    resourceModuleName: string,
    resource: ResourceModel,
  ) {
    this.context = context;
    this.moduleName = moduleName;
    this.resourceModuleName = resourceModuleName;
    this.resource = resource;
    this.resourceModuleAlias = this.context.project.getResourceModuleAlias(this.resource.type);
    this.sourceFile = this.context.project.createResourceTestFile(this.moduleName);
  }

  collectData(): void {
    this.context.logger.info(`Generating code for resource`);
  }

  generateCode(): void {
    this.context.logger.info(`Generating code for resource`);

    this.sourceFile.writeDefmodule(this.moduleName, () => {
      this.sourceFile.writeUse('ExUnit.Case', 'async: false');
      if (this.resource.jsonSchema()) {
        this.sourceFile.writeImport(this.context.testObjectFactoryModuleName);
      }
      this.sourceFile.writeAlias(this.resourceModuleName);

      if (this.resource.jsonSchema()) {
        this.generateTestBuildFunction();
      }
      this.generateTestLinkFunction();
    });
  }

  protected generateTestData(): Map<string, JSONValue> {
    faker.seed(132489235);

    const resourceSchema = this.resource.jsonSchema();
    assert(resourceSchema);

    const result = new Map<string, JSONValue>();

    result.set('id', faker.datatype.number({ min: 1, max: 1000 }));

    const attributesSchema = resourceSchema.getPropertyOrThrow('attributes');
    for (const [attrName, attrSchema] of attributesSchema.properties) {
      switch (attrSchema.type) {
        case 'boolean':
          result.set(attrName, faker.datatype.boolean());
          break;
        case 'integer': {
          let min: number | undefined;
          if (attrSchema.minimum != null) {
            min = attrSchema.minimum;
          } else if (attrSchema.exclusiveMinimum != null) {
            min = attrSchema.exclusiveMinimum + 1;
          }
          let max: number | undefined;
          if (attrSchema.maximum != null) {
            max = attrSchema.maximum;
          } else if (attrSchema.exclusiveMaximum != null) {
            max = attrSchema.exclusiveMaximum - 1;
          }
          result.set(attrName, faker.datatype.number({ min, max }));
          break;
        }
        case 'number': {
          let min: number | undefined;
          if (attrSchema.minimum != null) {
            min = attrSchema.minimum;
          } else if (attrSchema.exclusiveMinimum != null) {
            min = attrSchema.exclusiveMinimum + 1;
          }
          let max: number | undefined;
          if (attrSchema.maximum != null) {
            max = attrSchema.maximum;
          } else if (attrSchema.exclusiveMaximum != null) {
            max = attrSchema.exclusiveMaximum - 1;
          }
          result.set(attrName, faker.datatype.number({ min, max, precision: 0.01 }));
          break;
        }
        case 'string': {
          switch (attrSchema.format) {
            case 'password':
              result.set(attrName, faker.internet.password(8));
              break;
            case 'date': {
              const now = Date.now();
              const tenDays = 10 * 86400 * 1000;
              const date = faker.date.between(now - tenDays, now + tenDays);
              result.set(attrName, `"${date.toISOString().replace(/T.*$/g, '')}"`);
              break;
            }
            case 'date-time': {
              const now = Date.now();
              const tenDays = 10 * 86400 * 1000;
              const date = faker.date.between(now - tenDays, now + tenDays);
              result.set(attrName, `"${date.toISOString()}"`);
              break;
            }
            default: {
              const min = attrSchema.minLength ?? 0;
              const max = attrSchema.maxLength ?? 1024;
              result.set(attrName, `"${faker.lorem.word({ length: { min, max } })}"`);
              break;
            }
          }
          break;
        }
        default:
          assert.fail(
            `Unexpected schema type "${String(attrSchema.type)}" for attribute "${attrName}"`,
          );
      }
    }

    const relationshipsSchema = resourceSchema.getPropertyOrThrow('relationships');
    for (const [relName] of relationshipsSchema.properties) {
      const key = `${relName}Id`;
      result.set(key, faker.datatype.number({ min: 1, max: 1000 }));
    }

    return result;
  }

  protected generateTestBuildFunction(): void {
    const resourceSchema = this.resource.jsonSchema();
    assert(resourceSchema);

    this.sourceFile.newLine();
    this.sourceFile.writeLine('test "build/1" do');
    this.sourceFile.writeIndented(() => {
      // prepare part
      const values = this.generateTestData();

      this.sourceFile.writeLine('config = build(');
      this.sourceFile.writeIndented(() => {
        this.sourceFile.writeLine(`:${snakeCase(this.resource.type)},`);
        this.sourceFile.writeLine(`id: ${String(values.get('id'))},`);

        for (const attrName of this.resource.attributes.keys()) {
          const attrValue = values.get(attrName);
          const elixirAttrName = snakeCase(attrName);
          this.sourceFile.writeLine(`${elixirAttrName}: ${String(attrValue)},`);
        }
        for (const [relName, relObj] of this.resource.relationships) {
          const relValue = String(!!relObj.jsonSchema());
          const elixirRelName = snakeCase(`${relName}Id`);
          this.sourceFile.writeLine(`${elixirRelName}: ${String(relValue)},`);
        }
      });
      this.sourceFile.writeLine(')');
      this.sourceFile.newLine();

      // assert part
      this.sourceFile.writeLine(
        `assert ${this.resourceModuleAlias}.build(config) == %Jabbax.Document.Resource{`,
      );
      this.sourceFile.writeIndented(() => {
        this.sourceFile.writeLine(`type: "${this.resource.type}",`);
        this.sourceFile.writeLine(`id: ${String(values.get('id'))},`);

        if (this.resource.attributes.size) {
          this.sourceFile.writeLine('attributes: %{');
          this.sourceFile.writeIndented(() => {
            for (const attrName of this.resource.attributes.keys()) {
              const elixirAttrName = snakeCase(attrName);
              this.sourceFile.writeLine(`${elixirAttrName}: ${String(values.get(attrName))},`);
            }
          });
          this.sourceFile.writeLine('},');
        } else {
          this.sourceFile.writeLine('attributes: %{},');
        }

        if (this.resource.relationships.size) {
          this.sourceFile.writeLine('relationships: %{');
          this.sourceFile.writeIndented(() => {
            for (const [relName, { otherResourceType }] of this.resource.relationships) {
              const elixirAttrName = snakeCase(relName);
              this.sourceFile.writeLine(
                `${elixirAttrName}: %Jabbax.Document.ResourceId{id: config.${elixirAttrName}.id, type: "${otherResourceType}"},`,
              );
            }
          });
          this.sourceFile.writeLine('},');
        } else {
          this.sourceFile.writeLine('relationships: %{},');
        }
      });
      this.sourceFile.writeLine('}');
    });
    this.sourceFile.writeLine('end');
  }

  protected generateTestLinkFunction(): void {
    this.sourceFile.newLine();
    this.sourceFile.writeLine('test "link/1" do');
    this.sourceFile.writeIndented(() => {
      const value = 3525;
      this.sourceFile.writeLine(
        `assert ${this.resourceModuleAlias}.link(%{ id: ${value} }) == %Jabbax.Document.ResourceId{`,
      );
      this.sourceFile.writeIndentedLines(`type: "${this.resource.type}",`, `id: ${value},`);
      this.sourceFile.writeLine('}');
    });
    this.sourceFile.writeLine('end');
  }
}
