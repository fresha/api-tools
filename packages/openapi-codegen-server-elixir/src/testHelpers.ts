import { Project } from '@fresha/code-morph-ex';
import { createRegistry } from '@fresha/json-api-model';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTestContext as createBaseTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory } from '@fresha/openapi-model/build/3.0.3';

import { Generator } from './parts';

import type { Context } from './context';

export const createTestContext = (phoenixApp: string, rootDir = '/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createBaseTestContext(openapi, rootDir);

  const project = new Project({
    rootDir,
    phoenixApp,
    useInMemoryFileSystem: true,
  });

  return {
    ...base,
    testObjectFactoryModuleName: `${project.getAppModuleName()}.Factory`,
    openapi,
    project,
    registry: createRegistry(),
  };
};

export const createGenerator = (phoenixApp = 'api_tools_web'): Generator => {
  const context = createTestContext(phoenixApp, '/');
  const generator = new Generator(context);
  return generator;
};
