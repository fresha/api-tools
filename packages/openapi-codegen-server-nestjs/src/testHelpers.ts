import { Nullable } from '@fresha/api-tools-core';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory, SchemaModel } from '@fresha/openapi-model/build/3.0.3';

import { Context } from './context';
import { Generator } from './Generator';

export const createTestContext = (nestApp = 'web', outputPath = '/tmp/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createTSProjectTestContext(openapi, outputPath);

  base.project.createSourceFile(
    `${outputPath}/${nestApp}.module.ts`,
    `import { Module } from '@nestjs/common';
    import { AppController } from './app.controller';
    import { AppService } from './app.service';

    @Module({
      imports: [],
      controllers: [AppController],
      providers: [AppService],
    })
    export class AppModule {}`,
  );

  return {
    ...base,
    nestApp,
    addDTO: jest.fn(),
  };
};

export const createGenerator = (nestApp = 'app', rootDir = '/'): Generator => {
  const context = createTestContext(nestApp, rootDir);

  let generator: Generator;
  // eslint-disable-next-line prefer-const
  generator = new Generator({
    ...context,
    addDTO: (name: string, schema: Nullable<SchemaModel>) => generator.addDTO(name, schema),
  });

  return generator;
};
