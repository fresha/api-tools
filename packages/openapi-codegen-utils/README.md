# @fresha/openapi-codegen-utils

This package contains shared code used by OpenAPI code generators.

## Installation

```bash
$ npm install @fresha/openapi-codegen-utils
```
