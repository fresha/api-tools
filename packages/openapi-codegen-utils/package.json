{
  "name": "@fresha/openapi-codegen-utils",
  "version": "0.1.9",
  "description": "Shared utilites for OpenAPI codegens",
  "main": "build/index.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsc",
    "build:watch": "tsc --watch",
    "check": "run-s lint test",
    "check:fix": "run-s lint:fix test",
    "clean": "rimraf ./build",
    "eslint": "eslint ./src",
    "eslint:fix": "eslint ./src --fix",
    "lint": "run-s eslint typecheck",
    "lint:fix": "run-s eslint:fix typecheck",
    "prebuild": "npm run clean",
    "prebuild:watch": "npm run clean",
    "test": "jest",
    "typecheck": "tsc --noEmit"
  },
  "keywords": [
    "JSON:API"
  ],
  "author": "Fresha Engineering",
  "contributors": [
    {
      "name": "Andriy Mykulyak",
      "email": "andriy@fresha.com",
      "url": "https://github.com/mykulyak"
    }
  ],
  "maintainers": [
    {
      "name": "Andriy Mykulyak",
      "email": "andriy@fresha.com",
      "url": "https://github.com/mykulyak"
    }
  ],
  "license": "MIT",
  "devDependencies": {
    "@fresha/eslint-config": "^0.1.1",
    "@fresha/jest-config": "^0.1.1",
    "@fresha/typescript-config": "^0.1.1",
    "@types/pluralize": "^0.0.29",
    "typescript": "^4.7.2"
  },
  "dependencies": {
    "@fresha/api-tools-core": "^0.4.2",
    "@fresha/openapi-model": "^0.6.6",
    "pluralize": "^8.0.0",
    "ts-morph": "^16.0.0",
    "winston": "^3.8.2"
  },
  "gitHead": "5830273395fc060b19d06814b9396ce07eea778d"
}
