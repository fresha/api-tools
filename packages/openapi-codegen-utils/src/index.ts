export * from './assert';
export * from './context';
export * from './generator';
export * from './jsonapi';
export * from './logging';
export * from './openapi';
export * from './string';
export * from './yargs';
