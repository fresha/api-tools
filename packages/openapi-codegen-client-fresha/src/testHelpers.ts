// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { getAPIName, getRootUrlOrThrow } from '@fresha/openapi-codegen-utils';

import type { Context } from './context';
import type { OpenAPIModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel, rootDir = '/'): Context => {
  const base = createTSProjectTestContext(openapi, rootDir);

  let apiRootUrl: string;
  try {
    apiRootUrl = getRootUrlOrThrow(openapi);
  } catch (err) {
    apiRootUrl = 'API_URL';
  }

  return {
    ...base,
    apiName: getAPIName(openapi),
    apiRootUrl,
  };
};
