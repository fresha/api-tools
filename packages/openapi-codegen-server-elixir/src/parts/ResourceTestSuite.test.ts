import { faker } from '@faker-js/faker';
import { addResourceAttributes, setResourceSchema } from '@fresha/openapi-codegen-utils';

import { createTestContext } from '../testHelpers';

import { ResourceTestSuite } from './ResourceTestSuite';

test('simple resource', () => {
  const context = createTestContext('awesome_web');

  const schema = context.openapi.components.setSchema('TestResource');
  setResourceSchema(schema, 'resource-tests');
  addResourceAttributes(schema, {
    num1: {
      type: 'number',
      minimum: 10,
      exclusiveMinimum: true,
      maximum: 20,
      exclusiveMaximum: true,
    },
    num2: { type: 'number', minimum: 10, maximum: 20 },
    int1: {
      type: 'integer',
      minimum: 10,
      exclusiveMinimum: true,
      maximum: 20,
      exclusiveMaximum: true,
    },
    int2: { type: 'integer', minimum: 10, maximum: 20 },
  });

  const resourceSchema = context.registry.addResourceSchema(schema);

  faker.seed(38459284924);

  const generator = new ResourceTestSuite(
    context,
    'AwesomeWeb.TestResource',
    'AwesomeWeb.TestResource',
    resourceSchema,
  );
  generator.collectData();
  generator.generateCode();

  expect(generator.sourceFile.getText()).toMatchSnapshot();
});
