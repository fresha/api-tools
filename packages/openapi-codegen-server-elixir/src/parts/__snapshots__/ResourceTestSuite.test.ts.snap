// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple resource 1`] = `
"defmodule AwesomeWeb.TestResource do
  @moduledoc false

  use ExUnit.Case, async: false
  import AwesomeWeb.Factory
  alias AwesomeWeb.TestResource

  test "build/1" do
    config = build(
      :resource_tests,
      id: 884,
      num1: 11.75,
      num2: 11.6,
      int1: 16,
      int2: 10,
    )

    assert ResourceTestsResource.build(config) == %Jabbax.Document.Resource{
      type: "resource-tests",
      id: 884,
      attributes: %{
        num1: 11.75,
        num2: 11.6,
        int1: 16,
        int2: 10,
      },
      relationships: %{},
    }
  end

  test "link/1" do
    assert ResourceTestsResource.link(%{ id: 3525 }) == %Jabbax.Document.ResourceId{
      type: "resource-tests",
      id: 3525,
    }
  end
end"
`;
